posts/2024/2024_06_21_Inigo_week_4.rst

```diff
@@ -0,0 +1,53 @@
+Week 4 into GSoC 2024: Weight transfer experiments, hardships, and results!
+===========================================================================
+
+.. post:: June 21 2024
+    :author: Iñigo Tellaetxe
+    :tags: google
+    :category: gsoc
+
+
+
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
+
+Well, this week was really intense. I spent most of the time trying to transfer the weights from the pre-trained PyTorch model of the TractoInferno dataset to the Keras model.
+I must say that thanks to the reduced size of the AutoEncoder, it was feasible to do it layer by layer without going crazy. PyTorch uses a *channels first* convention, whereas TensorFlow uses *channels last*, what means that all the weights in the convolutional layers had to be transposed.
+This was the easiest part, as it was just a matter of using ``np.transpose``.
+After doing this, I set a common input for both networks, and I compared their outputs. As expected, they were not the same, and they were not even close enough.
+Thus, matching the behavior of the PyTorch model with the Keras implementation became my objective. To achieve so, I run a common input through all the layers of both models sequentially, and systematically compared the outputs of each layer.
+In the Encoder block, I found all the outputs to be within a reasonable range of each other (MAE = 1e-6), except for the last two operations, which flatten the output of the 1D convolutional layers and then feed it to a fully connected layer.
+This was partially good news, because most of the Encoder was behaving as desired, but, the most challenging part was adapting the flattening and reshaping operations happening in the encoder and the decoder, respectively.
+As the Keras 1D convolutional output dimensions do not follow the same ordering as in PyTorch, (*[n, m, channels]* vs *[n, channels, m]*), the flattening behavior of the models was different, and thus, the fully connected layer of the Encoder (named ``fc1``) was receiving different inputs.
```

Comment:
the flattening behavior of the models was different (the elements followed a different sorting when being concatenated into a 1D array)

If the above is accurate as to what was happening.

Reply:
Yes, correct. Added the comment in parenthesis, it clarifies it further. TY